File: python/dsa/algorithms/sliding-window/longest-substring.py
def longest_substring(s):
    char_map = {}
    left = 0
    max_len = 0

    for right in range(len(s)):
        # 1️⃣ Expand
        char = s[right]

        # 2️⃣ Condition: Check for duplicates
        if char in char_map:
            # 3️⃣ Contract: Jump left pointer
            if char_map[char] + 1 > left:
                left = char_map[char] + 1

        # 4️⃣ Record / Optimize
        char_map[char] = right
        if right - left + 1 > max_len:
            max_len = right - left + 1

    return max_len

File: javascript/dsa/algorithms/sliding-window/longest-substring.js
function longestSubstring(s) {
    let charMap = new Map();
    let left = 0;
    let maxLen = 0;

    for (let right = 0; right < s.length; right++) {
        // 1️⃣ Expand
        let char = s[right];

        // 2️⃣ Condition: Check for duplicates
        if (charMap.has(char)) {
            // 3️⃣ Contract: Jump left pointer
            if (charMap.get(char) + 1 > left) {
                left = charMap.get(char) + 1;
            }
        }

        // 4️⃣ Record / Optimize
        charMap.set(char, right);
        if (right - left + 1 > maxLen) {
            maxLen = right - left + 1;
        }
    }

    return maxLen;
}